### PR TITLE
Hang when reattaching after detach during shutdown

### LINKED ISSUE #6084: Fatal python error when detach in a rust thread during finalization
### Bug Description

I am examining a bug report in a PyO3 project - https://github.com/curioswitch/pyqwest/issues/164

It is essentially in the class of "abort instead of hang on finalization" bugs fixed broadly in Python 3.14 itself, with PyO3 fixing the issue on previous versions for `attach` in #4874, but there is still an issue with `detach`, if not when detaching at least when reattaching. In fact, Python 3.14 specifically added the hanging behavior for `PyEval_RestoreThread` as well (also `PyEval_AcquireThread` though I'm not sure if PyO3 uses that)

https://docs.python.org/3/c-api/threads.html#c.PyEval_RestoreThread

So in principle, I guess the exact same change needs to be done to thread restoration as well. 

~~Unfortunately, I was unable to create a reliable unit test for this. I could get this different error message, but I think it's a different issue.~~

I could come up with a test that fails reliably without the fix! Sent #6085 with the proposal

### Steps to Reproduce

Taken from original bug report - it fails reliably for me as well.

```
docker run --rm python:3.11-slim bash -c '
pip install -q pyqwest==0.6.0
cat > /tmp/r.py <<"PY"
import threading, time, pyqwest
def h(u, t):
    c = pyqwest.SyncClient()
    while True:
        try: c.get(u, timeout=t)
        except Exception: pass
for _ in range(16): threading.Thread(target=h, args=("http://127.0.0.1:9/", 1.0), daemon=True).start()
for _ in range(8):  threading.Thread(target=h, args=("http://192.0.2.1:81/", 5.0), daemon=True).start()
time.sleep(1.0)
PY
for i in $(seq 1 12); do python /tmp/r.py; echo "iter $i exit=$?"; done
'
```

### Backtrace

```shell
#2  abort ()
#4  __libc_fatal ()                                  "FATAL: exception not rethrown"
#6  __rustc::__rust_panic_cleanup ()
#7  std::panicking::catch_unwind::cleanup ()
#8  pyo3::impl_::trampoline::trampoline ()
#9  pyo3::impl_::trampoline::fastcall_cfunction_with_keywords ()
#10 method_vectorcall_FASTCALL_KEYWORDS ()
...
#22 thread_run () at ./Modules/_threadmodule.c:1124
#23 pythread_wrapper () at Python/thread_pthread.h:241
```

### Your operating system and version

Linux

### Your Python version (`python --version`)

Python 3.11

### Your Rust version (`rustc --version`)

1.95.0 (59807616e 2026-04-14)

### Your PyO3 version

0.28.3

### How did you install python? Did you use a virtualenv?

manylinux docker image

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by anuraaga**
Fixes #6084 

By submitting these contributions you agree for them to be dual-licensed under PyO3's [MIT OR Apache-2.0 license](https://github.com/PyO3/pyo3#license).

### CODE REVIEW

File: pyo3-ffi/src/ceval.rs [hunk 1]
```diff
@@ -139,6 +137,47 @@ extern_libpython! {
...
+struct HangThread;
+
+#[cfg(not(any(Py_3_14, target_arch = "wasm32")))]
+impl Drop for HangThread {
+    fn drop(&mut self) {
+        loop {
+            std::thread::park(); // Block forever.
+        }
+    }
+}
```
Comment (davidhewitt) [inline pyo3-ffi/src/ceval.rs]: Please move this to `pyo3-ffi/src/impl_/mod.rs` and deduplicate with the same logic in `pystate.rs`.

File: pytests/tests/test_detach_then_reattach_in_thread.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    reason="causes a crash on debug builds",
+)
+def test_detach_then_reattach_in_thread():
+    # A Python daemon thread calls block_in_detach_until_finalizing(), which releases
+    # the GIL via py.detach() and polls Py_IsFinalizing() (safe without GIL — reads
+    # an atomic). When the interpreter begins finalizing, the thread exits the closure
+    # and SuspendAttach::drop() calls PyEval_RestoreThread() on a finalizing interpreter.
+    t = threading.Thread(target=misc.block_in_detach_until_finalizing, daemon=True)
+    t.start()
+    # No join — thread polls until finalization
```
Comment (davidhewitt) [inline pytests/tests/test_detach_then_reattach_in_thread.py]: Can this please have a similar mechanism to `test_hammer_attaching_in_thread.py`, perhaps both files can be merged together as `test_finalization.py` or similar?

File: pytests/src/misc.rs [hunk 3]
```diff
@@ -31,6 +31,38 @@ fn hammer_attaching_in_thread() -> LockHolder {
...
+// Py_IsFinalizing() is public and exported on Python 3.13+ (where _Py_IsFinalizing was removed).
+// Both are safe to call without the GIL — they read an atomic variable.
+// PyO3 correctly only exports it on 3.13+ but for this unit test we can access the
+// older private version directly.
+fn is_finalizing() -> bool {
+    #[cfg(not(Py_3_13))]
+    {
+        extern "C" {
+            fn _Py_IsFinalizing() -> std::ffi::c_int;
+        }
```
Comment (davidhewitt) [inline pytests/src/misc.rs]: This will probably not work on PyPy or older x86 windows builds due to wrong symbol name. As per other comment I think this should be unified with `test_hammer_attaching_in_thread.py`

**Review by davidhewitt — COMMENTED** [review]
Thanks, in principle this seems ok.

General nit: a lot of the comments are very verbose, read like LLM-generated. If you did indeed use AI to help you produce this, please try to refrain from allowing the output to be too bloated.

File: pyo3-ffi/src/ceval.rs (hunk 1 above)
Comment (anuraaga) [reply pyo3-ffi/src/ceval.rs]: Thanks! I wasn't sure what a good place for the shared logic would be will do that

File: pytests/src/misc.rs (hunk 3 above)
Comment (anuraaga) [reply pytests/src/misc.rs]: I tried pretty hard to do that and it's in the first version of my issue comment. I couldn't avoid this different error

```
Fatal Python error: PyGILState_Release: thread state 0xefd390000910 must be current when releasing
Python runtime state: finalizing (tstate=0x000000000097a5d0)
```

I guess it could be about timing of being during GC or not. So I had to come up with a test case closer to the repro I got which was also polling until finalization. If you have any advice would be great

**Comment by anuraaga** [comment]
> If you did indeed use AI to help you produce this, please try to refrain from allowing the output to be too bloated.

Sorry about that - the logic change was quite easy and I could do that manually but did many iterations with AI to get a test case that flipped with the change. While I tried to clean them up probably some things are more obvious than I thought, will trim more.

File: pytests/src/misc.rs [hunk 4]
```diff
@@ -31,6 +31,34 @@ fn hammer_attaching_in_thread() -> LockHolder {
     LockHolder { sender }
 }
 
+/// Wrapper to mark Receiver as Sync.
+struct SyncReceiver<T>(std::sync::mpsc::Receiver<T>);
```
Comment (anuraaga) [inline pytests/src/misc.rs:35]: Incidentally I've always wondered if PyO3 could provide an easier way for this :-)

**Comment by anuraaga** [comment]
@davidhewitt I have gone ahead and merged the test with the existing pattern. Just as a reminder, the error message from this test is 

```
Fatal Python error: PyGILState_Release: thread state 0xefd390000910 must be current when releasing
Python runtime state: finalizing (tstate=0x000000000097a5d0)
```

while the one I was debugging is `"FATAL: exception not rethrown"`.

But since the same fix applied to both, maybe this new test is close enough in behavior that it is effectively a regression test for both, though I'm not sure. Since I've confirmed it fixes my issue I am comfortable with it

**Review by davidhewitt — APPROVED** [review]
I think the slight mismatch to your repro is ok; I think in both cases the cause is the `pthread_exit` jump causing issues during `PyEval_RestoreThread` call and patch here cleans this up.

Let's land this in 0.29. Thanks 👍

**Comment by anuraaga** [comment]
Sorry @davidhewitt I didn't realize clippy wasn't being run on PR, I added the safety note on the test code. I think the new failure is unrelated only with rust nightly

**Comment by davidhewitt** [comment]
> Sorry @davidhewitt I didn't realize clippy wasn't being run on PR, I added the safety note on the test code. I think the new failure is unrelated only with rust nightly

clippy is being run on PR but the difference is I merged #6093 to push forward a slow process to improve our internal quality.

**Comment by davidhewitt** [comment]
There is a genuine graalpy failure, I'm investigating...

**Comment by davidhewitt** [comment]
I decided to skip on graalpy for now; this is a compatibility issue in that implementation in my opinion. It'll presumably change with things like PEP 788 in the future.
